First pass at internationalizing amp-story

newmuis opened the pull request:
This PR adds a simple service that swaps strings out depending on the `lang` attribute of the nearest ancestor.  This service is currently in the `amp-story` extension, but could, in theory, be moved to `src/`.

The big problem with this is that it builds all strings for all languages into the binary that uses them.  This won't scale to a large number of strings or languages.  Ideally (I think?), we would have a way to build separate bundles based on the languages used, which we could dynamically fetch.  That becomes an interesting UX problem, since there could be a delay before we have the strings.

Fixes #13556.

newmuis commented:
/cc @calebcordry

gmajoulet commented on extensions/amp-story/0.1/messages.js:
```diff
@@ -0,0 +1,187 @@
+   *     document.  If unspecified, will use the document-level language, if
+   *     one exists, or the default otherwise.
+   */
+  getMessage(messageId, opt_elementToUse) {
```
Nit: the new styleguide asks for ``elementToUse = undefined`` instead

gmajoulet commented on src/services.js:
```diff
@@ -311,6 +311,24 @@ export class Services {
+   * @param {!Window} win
+   * @return {?Promise<?../extensions/amp-story/0.1/messages.MessageService>}
+   */
+  static messageServiceForOrNull(win) {
```
Since no other extension is planning on using this service, we have no risk of race conditions. Could we just get it from the synchronous ``messageService`` method?

calebcordry replied:
Don't forget about me 🙂

newmuis replied:
+1.  Also, this isn't stories-specific, so in theory, other people could use this for their strings too (/cc @cathyxz, does lightbox gallery have strings?)

cathyxz replied:
Nope, don't have strings on lightbox gallery (yet), but we may end up with aria strings in all components for accessibility.

alanorozco commented on extensions/amp-story/0.1/amp-story-desktop.css:
```diff
@@ -335,6 +334,7 @@ div.i-amphtml-story-top {
   margin: auto!important;
   color:#fff!important;
   box-sizing: initial !important;
+  text-transform: uppercase!important;
```
nit: space

alanorozco commented on extensions/amp-story/0.1/messages.js:
```diff
@@ -0,0 +1,187 @@
+   */
+  getLanguageCodesForElement_(element) {
+    const languageEl = closest(element, el => el.hasAttribute('lang'));
+    const languageCode = languageEl ? languageEl.getAttribute('lang') : null;
```
nit: you can probably fallback to empty string here.

newmuis replied:
`getAttribute` can return `null` if the attribute is not present, so I felt `null` was more consistent; languageCode is either the value of the `lang` attribute, or `null` if there wasn't one.

alanorozco replied:
sgtm

newmuis commented:
@choumx Done
@alanorozco I think you're right that's a better end goal, but we need a build step for that, and we don't currently have it.  This, at least, gets all the messages in a centralized place, so other refactors will be easier (i.e. they won't need to rip strings out that ended up in CSS somehow).  Also, if we allow multiple langs in the same doc, the build step becomes very complicated 😄

newmuis commented:
Fixed type errors, renamed "messages" (and all related names) to "localized strings", and added the ability to create pseudo-locales, which will help us out for testing.  We can expand pseudo-locale supoprt in a separate future PR, but this at least lets you verify that the string(s) you are using are indeed being passed through the `LocalizationService`, and are not hard-coded.

gmajoulet commented on extensions/amp-story/0.1/_locales/default.js:
```diff
@@ -0,0 +1,64 @@
+  },
+  [LocalizedStringId.AMP_STORY_SHARING_CLIPBOARD_FAILURE_TEXT]: {
+    string: ':(',
+    description: 'String shown in a toast to inform the user that a link ' +
```
Nit: a translator would probably not know what a toast is

newmuis replied:
Updated!

gmajoulet commented on extensions/amp-story/0.1/localization.js:
```diff
@@ -0,0 +1,231 @@
+ *
+ * Next ID: 22
+ *
+ * @const @enum {string}
```
Just curious: any reason why we want strings instead of numbers?

newmuis replied:
😦 

Yeah.  We directly use these as object keys in the `_locales/*.js` files, and object keys have to be strings.  If you let this be `@enum {number}` they get coerced to strings when used as keys, but then you get all kinds of nasty type errors when you try to use `Object.keys` because they're all strings, but anything that was accepting an ID is expecting a number.
